internal/cmd/beta/security-group-rule/create/create_test.go

```
@@ -0,0 +1,353 @@
+		},
+		Protocol: &iaas.CreateProtocol{
+			Int64:  utils.Ptr(int64(1)),
+			String: utils.Ptr("icmp"),
+		},
+		RemoteSecurityGroupId: utils.Ptr(testRemoteSecurityGroupId),
+	}
+	for _, mod := range mods {
+		mod(&payload)
+	}
+	return payload
+}
+
+func TestParseInput(t *testing.T) {
+	tests := []struct {
+		description   string
+		flagValues    map[string]string
+		isValid       bool
+		expectedModel *inputModel
+	}{
+		{
+			description: "base",
+			flagValues: fixtureFlagValues(func(flagValues map[string]string) {
+				delete(flagValues, portRangeMaxFlag)
+				delete(flagValues, portRangeMinFlag)
+				delete(flagValues, protocolNumberFlag)
+			}),
+			isValid: true,
+			expectedModel: fixtureInputModel(func(model *inputModel) {
+				model.PortRangeMax = nil
+				model.PortRangeMin = nil
+				model.ProtocolNumber = nil
+			}),
+		},
+		{
+			description: "required only",
+			flagValues: fixtureFlagValues(func(flagValues map[string]string) {
+				delete(flagValues, descriptionFlag)
+				delete(flagValues, etherTypeFlag)
+				delete(flagValues, icmpParameterCodeFlag)
+				delete(flagValues, icmpParameterTypeFlag)
+				delete(flagValues, ipRangeFlag)
+				delete(flagValues, portRangeMaxFlag)
+				delete(flagValues, portRangeMinFlag)
+				delete(flagValues, remoteSecurityGroupIdFlag)
+				delete(flagValues, protocolNumberFlag)
+				delete(flagValues, protocolNameFlag)
+			}),
+			isValid: true,
+			expectedModel: fixtureInputModel(func(model *inputModel) {
+				model.Description = nil
+				model.EtherType = nil
+				model.IcmpParameterCode = nil
+				model.IcmpParameterType = nil
+				model.IpRange = nil
+				model.PortRangeMax = nil
+				model.PortRangeMin = nil
+				model.RemoteSecurityGroupId = nil
+				model.ProtocolNumber = nil
+				model.ProtocolName = nil
+			}),
+		},
+		{
+			description: "direction missing",
+			flagValues: fixtureFlagValues(func(flagValues map[string]string) {
+				delete(flagValues, directionFlag)
+			}),
+			isValid: false,
```

If set to true, this test will fail because of `error validating flag groups: if any flags in the group [protocol-number protocol-name] are set none of the others can be; [protocol-name protocol-number]` but it should fail because of "direction missing" (false positive).

This is a general issue with the tests. I would like to have tests which expect e.g. that "direction is missing" and if something has been forgotten the test should fail. I will raise this issue in the team.

---

adapted unit test that, if isValid is set to true, it will fail with the correct error message